### PR TITLE
Refactor param automation mess

### PR DESCRIPTION
**Pull request by b-ma**
Hey @orottier,

Here is an attempt for cleaning the huge messy `match` in `AudioParamProcessor::compute_buffer` (which has been boring me since a long time...). It's not finished at all, I just rewrote the `SetValue` stuff for now, I prefer to have your feedback before spending more time on it.

What do you think, should I go on?

(Maybe that could help also with perf by minimizing cache, I don't really know)

### CODE REVIEW

File: src/param.rs [hunk 1]
```diff
@@ -1075,22 +1131,34 @@ impl AudioParamProcessor {
...
             self.buffer.push(self.intrinsic_value);
             // nothing to compute in timeline, for both k-rate and a-rate
             if is_constant_block {
                 return;
             }
         }
 
+        // For borrow check reasons, we move move out the `event_timeline`,
+        // and pass it around as &mut.
+        let mut event_timeline = std::mem::take(&mut self.event_timeline);
```
Comment (b-ma) [inline src/param.rs:1144]: Just used the same strategy as in graph render to workaround the borrow checker. Maybe there is a more simple solution but I couldn't find it...

Comment (b-ma) [reply src/param.rs:1144]: Also, `EventTimeline` contains a `Vec`, do it allocates when replacing with default?

Comment (b-ma) [reply src/param.rs:1144]: Hum, I guess this is my answer: https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-Default-for-Vec%3CT,+Global%3E

**Comment by orottier** [comment]
Hey, definitely anything that reduces the `match` method complexity is fine with me.

We are maybe able to make the pattern you are suggesting a bit more elegant. I don't think you need to take out the `&mut event_timeline` to work around the borrow checker

If you make the EventType a Copy-type:
```
#[derive(PartialEq, Eq, Debug, Copy, Clone)]
enum AudioParamEventType {
```

And start the big match with
```
        loop {
            let next_event_type = self.event_timeline.peek().map(|e| e.event_type);

            match next_event_type {
                None => {
// snip
                    break;
                }
                Some(AudioParamEventType::SetValue) | Some(AudioParamEventType::SetValueAtTime) => {
                    if self.compute_set_value_automation(&block_infos) {
                        break;
                    }
                }
               // etc
```

A drawback with your suggested solution is that we `peek` twice for every handled event, but maybe that is acceptable

**Comment by b-ma** [comment]
Ok thanks, your method worked like a charm!

Pretty happy this is cleaned up :)

**Comment by orottier** [comment]
Tests pass, code is more readable. Great work

**Pull request merged**